### ISSUE TITLE
Spurious non-allocatedness in old state error in twostate predicate

### ISSUE THREAD

**Issue by ssomayyajula** [dafny-lang/dafny#5722]
### Dafny version

master

### Code to produce this issue

```dafny
class Member
{
  var b: bool
  
  // Ok!
  twostate predicate I()
    reads this
  {
    old(b) && b
  }
}

class Composite
{
  var mem: Member

  constructor()
  {
    mem := new Member;
  }

  // Error: mem could not be proved to be allocated in Member.I()'s previous state
  twostate predicate I()
    // Line below doesn't help either
    //requires old(allocated(mem))
    reads this, mem
  {
    mem.I()
  }
}
```

### Command to run and resulting output

```code
Error: receiver argument could not be proved to be allocated in the two-state function's previous state: mem.I()
```

### What happened?

Any valid invocation of `c.I@OLD()` (i.e., when `c: Composite` is provably allocated at `OLD`) will necessarily imply, at that point, that `old@OLD(allocated(c.mem))`. Yet, `Composite.I()` does not verify, even if `old(allocated(mem))` is explicitly assumed.

### What type of operating system are you experiencing the problem on?

Mac

**Comment by RustanLeino** [comment]
Dafny complains that the value of `mem` in the body of `Composite.I()` may not have been allocated in the "old" state on which `Member.I()` is being invoked. The complaint is valid, since the specification of `Composite.I()` allows a caller

```
method Caller(c: Composite)
  modifies c
{
  var m := new Member;
  c.mem := m;
  var i := c.I();
}
```

To write the body of `Composite.I()` given above, the two-state predicate `Composite.I()` needs some precondition. The precondition you attempted is

```
requires old(allocated(this.mem))
```

where I made the "`this.`" explicit. However, as Dafny points out, this specification is not what you need. To understand what's going on, note that the heap is used in _two_ places in the expression `allocated(this.mem)`. One of the two places is the `.` in `this.mem` and the other is the `allocated(_)`. The enclosing use of `old` is making both of them refer to the old heap. 

In other words, the attempted precondition says that if you evaluate `this.mem` _in the old state of `Composite.I()`_ , then you get a value that was allocated already in the old state of `Composite.I()`. But the body of `Composite.I()` uses the _current_ value of `this.mem`, not the old value of `this.mem`. So, what you want to say is: if you evaluate `this.mem` in the _current_ state, then you get a value that was allocated already in the old state. That is, you must arrange for `this.mem` to be evaluated outside the `old` and then applied `old(allocated(_))` to it. Here's how you do that:

```
requires var currentMem := mem; old(allocated(currentMem))
```
